File: src/nodes/data/bool.rs
use apica_common::bytecodes::ApicaSpecificationBytecode;
use crate::utils::position::Position;

pub struct NodeDataBool {
    position: Position,
    bytecode: ApicaSpecificationBytecode,
    value: bool,
}

impl NodeDataBool {
    pub fn init(position: Position, bytecode: ApicaSpecificationBytecode, value: bool) -> NodeDataBool {
        NodeDataBool { position, bytecode, value }
    }
    
    pub fn get_position(&self) -> &Position {
        &self.position
    }
    
    pub fn show(&self, indent: &mut String, end: char) {
        print!("{indent}NodeDataBool(data: {:?}, value: {}){end}", self.bytecode, self.value);
    }
    
    pub fn get_bytecode(&self) -> ApicaSpecificationBytecode {
        self.bytecode
    }
    
    pub fn get_value(&self) -> bool {
        self.value
    }
}

File: src/nodes/data/bool.cpp
#include "nodes/data/bool.hpp"
#include "core/emitter.hpp"
#include <iostream>

using namespace nodes;

NodeDataBool::NodeDataBool(const utils::Position &position, common::bytecodes::ApicaSpecificationBytecode spec_bytecode, bool value)
    : NodeData(position, spec_bytecode), value(value) {

}

void NodeDataBool::show(std::string &indent, char end) const {
    std::cout << indent << "NodeDataBool(spec: " << this->spec_bytecode << ", val: " << this->value << ')' << end;
}

NodeKind NodeDataBool::getKind() const {
    return NodeKind::DataBool;
}

void NodeDataBool::emit(core::Emitter &emitter) const {
    emitter.writeU64(this->spec_bytecode);
    emitter.writeU8(this->value);
}

bool NodeDataBool::getSpecificationValue() const {
    return this->value;
}